lib/ASTGen/Sources/ASTGen/Decls.swift

```diff
@@ -1,162 +1,503 @@
 import CASTBridging
-import SwiftParser
 import SwiftSyntax
 
+// MARK: - TypeDecl
+
 extension ASTGenVisitor {
   public func visit(_ node: TypealiasDeclSyntax) -> ASTNode {
-    let aliasLoc = bridgedSourceLoc(for: node.typealiasKeyword)
-    let equalLoc = bridgedSourceLoc(for: node.initializer.equal)
-    var nameText = node.identifier.text
-    let name = nameText.withBridgedString { bridgedName in
-      return ASTContext_getIdentifier(ctx, bridgedName)
-    }
-    let nameLoc = bridgedSourceLoc(for: node.identifier)
-    let genericParams = node.genericParameterClause.map { self.visit($0).rawValue }
-    let out = TypeAliasDecl_create(
-      self.ctx, self.declContext, aliasLoc, equalLoc, name, nameLoc, genericParams)
+    let (name, nameLoc) = self.bridgedIdentifierAndSourceLoc(for: node.identifier)
 
-    let oldDeclContext = declContext
-    declContext = out.declContext
-    defer { declContext = oldDeclContext }
+    return .decl(
+      TypeAliasDecl_create(
+        self.ctx,
+        self.declContext,
+        self.bridgedSourceLoc(for: node.typealiasKeyword),
+        name,
+        nameLoc,
+        node.genericParameterClause.map(self.visit)?.rawValue,
+        self.bridgedSourceLoc(for: node.initializer.equal),
+        self.visit(node.initializer.value).rawValue,
+        node.genericWhereClause.map(self.visit)?.rawValue
+      )
+    )
+  }
 
-    let underlying = self.visit(node.initializer.value).rawValue
-    TypeAliasDecl_setUnderlyingTypeRepr(out.nominalDecl, underlying)
+  public func visit(_ node: EnumDeclSyntax) -> ASTNode {
+    let (name, nameLoc) = self.bridgedIdentifierAndSourceLoc(for: node.identifier)
 
-    return .decl(out.decl)
+    let declAndDeclContext = (node.inheritanceClause?.inheritedTypeCollection).map {
+      self.visit($0.typeName).rawValue
+    } withBridgedArrayRef: { inheritedTypes in
+      EnumDecl_create(
+        self.ctx,
+        self.declContext,
+        self.bridgedSourceLoc(for: node.enumKeyword),
+        name,
+        nameLoc,
+        node.genericParameterClause.map(self.visit)?.rawValue,
+        inheritedTypes,
+        node.genericWhereClause.map(self.visit)?.rawValue
+      )
+    }
```

Comment:
I think it's a great idea to use BumpPtrAllocator.

> we should maybe not only consider an optimization for the case that `count == 1`

I wish we can still have a optimization for `count == 1` case (i.e. not allocating any heap memory,) but I don't think we have a "safe" way to do it. Allocating one element sounds like a waste, but it's a reasonable trade-off.

@ahoppen what was your idea to get a bridged array from `OftenSmallArray<Element>.one(element)` case without using `withXXX()` pattern?

> We could even consider resetting the bump allocator at certain points.

I don't think we need this. And the lifetime of `ASTGenVisitor` is only for creating a AST for each SourceFile, and the memory allocation would always be less than the result AST. So I don't think the allocator would waste much memory for unreasonably long time.

Reply:
> @ahoppen what was your idea to get a bridged array from `OftenSmallArray<Element>.one(element)` case without using `withXXX()` pattern?

I don’t have an idea, probably should have mentioned that. But I thought that we could have a function that gets bridged arrays for 1, 2, 3, …  `OftenSmallArray`s so that this would read something like the following, which I think would be an improvement. With `OftenSmallArray`, this would read as follows:

```swift
let inheritedTypes = node.inheritanceClause?.inheritedTypeCollection.map(in: self) {
  self.visit($0.typeName).rawValue
} ?? []
return withBridgedArray(inheritedTypes, /* could also take 2 arguments */) { inheritedTypes in
  return EnumDecl_create(…)
}
```

> I don't think we need this. The lifetime of `ASTGenVisitor` is only for creating a AST for each SourceFile, and the memory allocation would always be less than the result AST. So I don't think the allocator would waste much memory for unreasonably long time.

The main point I wanted to make here was that *if* we discover that the bump allocator uses measurable memory we *could* make such an optimization, which should significantly reduce the memory overhead of it even further. 

I also realized that I didn’t mention that we do have a `BumpPtrAllocator` implementation in swift-syntax that can be used here: https://github.com/apple/swift-syntax/blob/main/Sources/SwiftSyntax/BumpPtrAllocator.swift

Reply:
>  I thought that we could have a function that gets bridged arrays for 1, 2, 3, …  OftenSmallArrays

I see, that way we can avoid nested `withBridgedArray` call, yes that's an improvement. But I don't think it's worth it. Allocating one pointer in `BumpPtrAllocator` is not that heavy, readability is more important. :+1:

Reply:
I went with bump allocation as suggested. The sequence transforming methods I had to overload are `map`, `compactMap`, and `reversed`, which I’m frankly not too happy about. I wish we could get away with extending `LazySequenceProtocol` with a `toBridgedArrayRef(in:)` method instead. The problem is we need to at least know the max count for allocation, but there is no such API and getting the `count` on lazy wrappers can be painfully expensive.